### PR TITLE
Postgres v0.1

### PR DESCRIPTION
**Pull request by byucesoy**
This is the initial commit for PostgreSQL service. It contains bare minimums
to introduce PostgreSQL service. Currently, we only create a virtual machine,
install PostgreSQL binaries in it and do some basic configurations. In the
upcoming commits we will add things like UI, API support, HA, backup/restore,
firewall rules, DNS names, scaling up/down, certificates and many more. Even
though this is incomplete, it provides enough infrastructure to build on top
of it.

Here are few notable design decisions we made in this commit;
- We don't ask for database name, username or password from user. They will
all be created by default. Database name and username is postgres, password
will be auto generated at run-time.
- Corollary of above, we are giving superuser access to users. This might be
biggest divergence of our postgres service implementation from the incumbents
but it is not an unseen practice. There are numerous benefits of this; most
importantly, it gives more power to users, which could tilt decision to our
favor for users who want or need superuser access. Secondly, it makes our
implementation simpler; we don't need to implement special logic to support
role or extension creation. Users can do these things themselves without
needing us to provide special API/UI. On the flip side, it is possible to
shoot yourself on the foot with too much power, and thus create additional
support burden. I expect this to be rare occasion, but we might want to
consider our support model for such cases.
- We are also introducing a new concept, that is not just about postgres in
this commit; service projects. PostgreSQL service only exposes PostgreSQL
resource to user not VMs or public IP addresses it creates in the back end.
Those resources needs to live in somewhere. Service project is exactly that.
Each service will have their own account and project where they can keep
the backing resources. In this commit, we are creating one project for all
PostgreSQL services' resources, but after gathering more data, we might
change that in the future.

Currently there is no API or GUI interface. To be able to create PostgreSQL
resources, you need to set `ENV["POSTGRES_SERVICE_PROJECT_ID"]` in your .env.rb.
    
Then you can create new postgres database with this command;
```
Prog::Postgres::PostgresNexus.assemble(<project-id>, <location>, "test-pg-server", "standard-4", 100)
```
After it is ready, you can get the connection string like this;
```
Postgres::PostgresServer.first.connection_string
```

### CODE REVIEW

File: ubid.rb [hunk 1]
```diff
@@ -55,6 +55,7 @@ class UBID
   TYPE_PAYMENT_METHOD = "pm"
   TYPE_GITHUB_INSTALLATION = "g1"
   TYPE_GITHUB_RUNNER = "gr"
+  TYPE_POSTGRES_SERVER = "pd"
```
Comment (fdr) [inline ubid.rb:58]: Oh noooooo `pd` because `pg` is taken? Man, I wonder if we should take this one in the teeth and reassign Page to a new shortcode given they're not really external identifiers and we're pretty early in this whole thing.

File: migrate/20230915_postgres.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      column :target_vm_size, :text, collate: '"C"', null: false
+      column :target_storage_size_gib, :bigint, null: false
+      column :superuser_password, :text, collate: '"C"', null: false # TODO encrypt this
+      column :vm_id, :vm, type: :uuid
+    end
+
+    alter_table(:billing_record) do
+      drop_constraint(:billing_record_resource_id_span_excl)
+      add_exclusion_constraint([[:resource_id, "="], [:billing_rate_id, "="], [:span, "&&"]])
+    end
```
Comment (fdr) [inline migrate/20230915_postgres.rb:21]: This should be good to chip out into another patch, which would probably simplify this patch to a `change` block and get the new constraint tested, since it seems like this is more a change to how billing constraints work in general...and perhaps an oversight in the original?

File: ubid.rb (hunk 1 above)
Comment (byucesoy) [reply ubid.rb:58]: Yep. After seeing pg is taken, I attempted to use ps as postgres server, but ps is also taken (private subnet) :smiling_face_with_tear:. Maybe this was just bad luck, but getting 2 back-to-back conflicts this soon in our life cycle made me think that 2 characters might not be enough as add more services/models.

I'm totally onboard with stealing pg from pages. I'll extract this to separate PR so that we can deploy it sooner with more control.

Comment (fdr) [reply ubid.rb:58]: I guess I was originally inspired by the FEBE postgres protocol, where *one* character was enough. My guess is "s" is pretty heavily subscribed (server) or certain other characters.

We might have to be a bit more vigilant when making these up

File: migrate/20230915_postgres.rb (hunk 2 above)
Comment (byucesoy) [reply migrate/20230915_postgres.rb]: It was oversight. I thought about this case, but forgot about it during implementation. I extracted this to separate PR and merged it.

File: ubid.rb (hunk 1 above)
Comment (byucesoy) [reply ubid.rb]: I committed and deployed a change to make page shortcode to be pa.

File: rhizome/common/lib/util.rb [hunk 3]
```diff
@@ -65,3 +66,9 @@ def safe_write_to_file(filename, content)
     File.rename(temp_filename, filename)
   end
 end
+
+def change_os_user(user)
```
Comment (fdr) [inline rhizome/common/lib/util.rb:70]: I kinda wonder if this is a good idea in this situation relative to `sudo -u` or `systemd-run --uid`. I have written such programs before, but only when transitioning in the middle of a program is necessary, whereas here, it seems like it runs in the beginning only?

File: migrate/20230920_postgres.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    create_table(:postgres_server) do
+      column :id, :uuid, primary_key: true, default: nil
+      column :created_at, :timestamptz, null: false, default: Sequel.lit("now()")
+      column :updated_at, :timestamptz, null: false, default: Sequel.lit("now()")
+      column :project_id, :project, type: :uuid, null: false
```
Comment (fdr) [inline migrate/20230920_postgres.rb:9]: I think this is suspicious given the authorization model, although I can appreciate the convenience. Would a method do the job? technically more than one project can associate with a resource, though we don't use that feature for anything at present (it may be useful for transitioning from one set of tags and policies to another in a smooth way)

**Comment by fdr** [comment]
I think one thing left out of the commit message, which is how to work with this. It lacks both GUIs and API interfaces, plus it'd be nice to have a prescription from the designer how to get working with it in any case. Some of the existing content could be dropped in preference for this more useful information (for example, ", but after gathering more data, we might
change that in the future." probably exists at the margin of utility).

I have found LLM programs useful for suggestion of what to cut, although it's kinda laborious and slow to use them. Or maybe it can just be long, it's the first composed service we're putting into circulation after the much-older #155, after all.

File: migrate/20230920_postgres.rb (hunk 4 above)
Comment (byucesoy) [reply migrate/20230920_postgres.rb:9]: @enescakir also raised this. Maybe this is good time to discuss N-to-N relationship of projects and resources.

I'm not entirely convinced on the benefit of being able to associate one resource to multiple projects at the cost of additional complexity. When we associate resources with multiple projects, we bring resource to the same level in the hierarchy as projects and accounts, and our model becomes more like graph than tree, which is more difficult to deal with.

On more practical side;
- Which project should be charged for a particular resource? For VMs we do projects.first, which is a bad hack IMO, but works as long as we don't introduce second project to the mix.
- With N-to-N association, deleting a project becomes a reference counting for each resource in it, which is doable, but why bother?
- Will we have no-access rules in our authorization model? I think we should, they are quite useful. However they would cause quite interesting problems with N-to-N relationship of projects and resources. Particularly, what do we do if a resource is in 2 projects and user defines access rule for one and no-access rule for the other.

Personally, I'd argue adding project_id to VMs as well. We can still keep N-to-N relationship between projects and accounts. Also users can still define access rules for individual resources.

File: rhizome/common/lib/util.rb (hunk 3 above)
Comment (byucesoy) [reply rhizome/common/lib/util.rb]: Removed this for now. We can add it back if we need it.

**Review by fdr — APPROVED** [review]
Well let's get it in to avoid conflicts, I don't see a problem with doing so given it's still quite malleable, and non-invasive.

File: migrate/20230920_postgres.rb (hunk 4 above)
Comment (fdr) [reply migrate/20230920_postgres.rb:9]: The only use case I have thought of for multi-project is transitioning from one security regime to another safely, being able to test it in phases. As it stands in industry today, this is an extremely narrow use case, just about everyone makes reorganizing the security apparatus for resources a high-stakes maneuver that cannot be tested, and everyone accepts it. This may be an acceptable thing to likewise inflict on people.

To enable two-phase opens non-trivial issues, like possibly having a secondary project that's insecure accessing a resource.

The era of developing this was quite a while ago, so perhaps our consideration was more favorable to experiment than tradition at that time. And since then, some things did come out, such as Enes's `Authorization` module. But in the current era it'd be more tempting to see if we could pare things back since we haven't made strides here in some time.

Comment (furkansahin) [reply migrate/20230920_postgres.rb:9]: This discussion also makes me think about the private subnets. What if customer wants to be able to connect 2 projects? Having 1 subnet listed under 2 projects make this much easier.

File: config/billing_rates.yml [hunk 5]
```diff
@@ -1,6 +1,12 @@
...
+- { id: 139d9a67-8182-8578-a303-235cabd5161c, resource_type: VmCores,         resource_family: standard, location: hetzner-fsn1,   unit_price: 0.000622768 }
+- { id: 08c502f7-df5d-8978-9896-feafa0ec5c40, resource_type: VmCores,         resource_family: standard, location: hetzner-hel1,   unit_price: 0.000595982 }
+- { id: 2acc2bbf-d451-49c6-93de-db8d15311042, resource_type: VmCores,         resource_family: standard, location: github-runners, unit_price: 0.000595982 }
+- { id: 118b7e2d-fa8d-8d78-910d-1f62fcc657ec, resource_type: IPAddress,       resource_family: IPv4,     location: hetzner-fsn1,   unit_price: 0.000074405 }
+- { id: 1bde2200-545a-8d78-960e-08a303111e3d, resource_type: IPAddress,       resource_family: IPv4,     location: hetzner-hel1,   unit_price: 0.000074405 }
+- { id: 38685ec4-38eb-4460-aed3-8a36bab14175, resource_type: IPAddress,       resource_family: IPv4,     location: github-runners, unit_price: 0.000074405 }
+- { id: 36b3883a-a72a-4c08-a4c6-482f4f91e148, resource_type: PostgresCores,   resource_family: standard, location: hetzner-fsn1,   unit_price: 0.001245536 }
```
Comment (furkansahin) [inline config/billing_rates.yml:7]: It would be good to explain why our postgresCores are the double of regular cores.

Comment (furkansahin) [reply config/billing_rates.yml:7]: you can do this in the commit message, too.

File: migrate/20230920_postgres.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      column :location, :text, collate: '"C"', null: false
+      column :server_name, :text, collate: '"C"', null: false, unique: true
+      column :target_vm_size, :text, collate: '"C"', null: false
+      column :target_storage_size_gib, :bigint, null: false
```
Comment (furkansahin) [inline migrate/20230920_postgres.rb:13]: I guess I understand why you want to keep target_vm_size and target_storage_size_gib here. This is a preparation for future when we implement scale-up? If so, can we leave a comment here about it. I would ideally ask you to leave it out of this patch but it would require a custom migration script when we introduce this capability to migrate the system. 

On the other hand, I'm not sure how soon we'll work on this. I leave the decision to you.

File: rhizome/postgres/bin/configure [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+# Refer to the "Client Authentication" section in the PostgreSQL
+# documentation for a complete description of this file.
+
+# TYPE  DATABASE        USER            ADDRESS                 METHOD
+# Database administrative login by Unix domain socket
+local   all             postgres                                peer
+
+# "local" is for Unix domain socket connections only
+local   all             all                                     peer
+# IPv4 local connections:
```
Comment (furkansahin) [inline rhizome/postgres/bin/configure:27]: maybe we can add the private subnet range here to allow private connections?

File: model/postgres/postgres_server.rb [hunk 8]
```diff
@@ -1,4 +1,60 @@
...
+  include Authorization::HyperTagMethods
+  include Authorization::TaggableMethods
+
+  semaphore :restart, :destroy
+
+  def hyper_tag_name(project)
+    "project/#{project.ubid}/location/#{location}/postgres/#{server_name}"
+  end
+
+  def vm
```
Comment (enescakir) [inline model/postgres/postgres_server.rb:22]: Since you have the following, no need to this helper
```
 one_to_one :vm
```

File: prog/postgres/postgres_nexus.rb [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+require_relative "../../lib/util"
+
+class Prog::Postgres::PostgresNexus < Prog::Base
+  extend Forwardable
+  def_delegators :server, :vm
+
+  semaphore :restart, :destroy
+
+  def self.assemble(project_id, location, server_name, vm_size, storage_size_gib)
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:13]: The `assemble` method for other resources use named parameters. Should we keep this convention?

```
self.assemble(public_key, project_id, name: nil, size: "standard-2",
    unix_user: "ubi", location: "hetzner-hel1", boot_image: "ubuntu-jammy",
    private_subnet_id: nil, nic_id: nil, storage_volumes: nil, boot_disk_index: 0,
    enable_ip4: false)
```

```
  def self.assemble(sshable_hostname, location: "hetzner-hel1", net6: nil, ndp_needed: false, provider: nil, hetzner_server_identifier: nil)
```

```
  def self.assemble(project_id, name: nil, location: "hetzner-hel1", ipv6_range: nil, ipv4_range: nil)
```

File: prog/postgres/postgres_nexus.rb [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    donate
+  end
+
+  label def mount_data_disk
+    case vm.sshable.cmd("common/bin/daemonizer --check format_disk")
+    when "Succeeded"
+      vm.sshable.cmd("sudo mkdir -p /dat")
+      device_path = vm.vm_storage_volumes.find { _1.boot == false }.device_path.shellescape
+
+      vm.sshable.cmd("sudo mount #{device_path} /dat")
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:95]: Should we add `fstab` record for it? As I know mount not persistent, so it will not mount after a reboot.

File: prog/postgres/postgres_nexus.rb [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+    nap 5
+  end
+
+  label def create_billing_record
+    BillingRecord.create_with_id(
+      project_id: server.project_id,
+      resource_id: server.id,
+      resource_name: server.server_name,
+      billing_rate_id: BillingRate.from_resource_properties("PostgresCores", "standard", server.location)["id"],
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:134]: Should we use `vm.family`?
```suggestion
      billing_rate_id: BillingRate.from_resource_properties("PostgresCores", vm.family, server.location)["id"],
```

File: prog/postgres/postgres_nexus.rb [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+    server.dissociate_with_project(server.project)
+    server.destroy
+
+    pop "postgres server is deleted"
+  end
+
+  def server
+    @server ||= PostgresServer[strand.id]
+  end
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:183]: You can use `subject_is` helper. If you want to use `server` instead of `postgres_server`, we may need to improve `subject_is` functionality.
```
subject_is: postgres_server
```

File: prog/postgres/postgres_nexus.rb [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+  label def destroy
+    register_deadline(nil, 5 * 60)
+    reap
+
+    decr_destroy
+
+    if vm
+      vm.sshable&.destroy
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:170]: Should we destroy the private subnet? Since it's in the service project, the user can't destroy it.

File: prog/postgres/postgres_nexus.rb [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+      vm.incr_destroy
+      donate
```
Comment (enescakir) [inline prog/postgres/postgres_nexus.rb:172]: Why does the strand donate here?

File: config/billing_rates.yml (hunk 5 above)
Comment (byucesoy) [reply config/billing_rates.yml:7]: We already have an explanation in the commit message of the related commit (the PR has 3 commits). Do you want it to be more detailed?

File: migrate/20230920_postgres.rb (hunk 6 above)
Comment (byucesoy) [reply migrate/20230920_postgres.rb:13]: We plan to look into this immediately after we have backup because backups are pre-requisite for failover based scale-up/down.

File: rhizome/postgres/bin/configure (hunk 7 above)
Comment (byucesoy) [reply rhizome/postgres/bin/configure:27]: Good idea, I'll ping you about this.

File: prog/postgres/postgres_nexus.rb (hunk 11 above)
Comment (byucesoy) [reply prog/postgres/postgres_nexus.rb:134]: Not necessarily, this is the family of the postgres service, which would usually be same with vm.family, but not always. For example we might use standard VMs for a business critical postgres family. I don't want to couple vm family and postgres family.

File: prog/postgres/postgres_nexus.rb (hunk 14 above)
Comment (byucesoy) [reply prog/postgres/postgres_nexus.rb:172]: It used to have child strands in a previous design, changed it to nap

File: prog/postgres/postgres_nexus.rb (hunk 9 above)
Comment (byucesoy) [reply prog/postgres/postgres_nexus.rb:13]: It seems we use named parameters only for optional parameters, other assemble methods don't use named parameters for non optional parameters. None of the parameters in postgres_nexus assemble is optional.

I'm fine with making all parameters(optional/non-optional) named, but that should be a different PR probably.

File: prog/postgres/postgres_nexus.rb (hunk 10 above)
Comment (byucesoy) [reply prog/postgres/postgres_nexus.rb:95]: This was in my notes, but then I forgot about it. Thanks for catching it. I'll look into it.

Comment (byucesoy) [reply prog/postgres/postgres_nexus.rb:98]: Done. I added a utility to safely adding records to fstab https://github.com/ubicloud/ubicloud/pull/578/commits/4904a4709e4ea9f05bd402db95759ef3a79de764

File: prog/postgres/postgres_nexus.rb (hunk 9 above)
Comment (enescakir) [reply prog/postgres/postgres_nexus.rb:15]: Makes sense

File: migrate/20230920_postgres.rb (hunk 4 above)
Comment (byucesoy) [reply migrate/20230920_postgres.rb:9]: We talked about this offline and decided to collect more data points to before making a decision on n-to-n relationships. For now, I kept the project_id. I'll also add a DB constraint to prevent potential unintentional disagreements between two models.

**Comment by fdr** [comment]
>  Enable private connections for PostgreSQL

> With this commit, we add private IPv4 and IPv6 ranges to the pg_hba.conf, thus
allowing connections from private subnet, which is should be faster as we don't
need to go out to public internet.
> In practice, I didn't observe huge difference, because in one example I tried,
the route used by public communication was also pretty good. However, it is
possible to pick worse paths for public communication.

This is a pretty complicated topic. There's an immediate problem with the justification as given though: most providers that only give us public internet IP addresses, and then we add tunneling. If anything, our overlay private networking *adds* overhead (encryption and encapsulation), in normal operation. We don't have any kind of internal "fast path" that avoids an edge-type device, as some (all?) of the hyperscalers tend to have.

There is an asterisk here: the situation gets more complicated once we add the "floating" IP address features some providers have. *These* may have a performance differential with the more basic IPv6 ephemeral/host-linked IP addresses we use for the IPsec underlay. The routers used to serve that traffic may be different or the same, more, or less, congested on average. I think, in general, that the more restricted host-linked addresses will tend to be faster, and so here we might secure some kind of advantage. But I think we'll be losing on average, so really our justifications lie elsewhere.

I think these are the two fundamental reasons:

* We'd like to shield Postgres from the public Internet.
* We'd want a layer of indirection as it comes to addressing for a given Postgres server.

Let's start with the first one, because it justifies this feature in a clear way:

A Postgres server with the normal authentication overhead, even for a failed authentication, involves a fork, i.e. it's trivial to fork-bomb a Postgres via FEBE over the Internet. IPsec is significantly harder to attack in this way. Postgres is not alone in not being robust to DOS attacks from bad clients, so having this feature in general is desirable.

The second is also pretty important, but it has a lot of ramifications:

* Flexibility over the address also makes firewalls more usable for the customer, since they are reliant on pattern matching IPs, and at least in IPv4, the ranges tend to be fragmented, expensive, too small, or all of the above.

* We can always offer a dual-stacked floating-IP like situation inside networks that is portable/constant across platforms (down to the quirks in the mechanism) inside an encapsulation (by reprogramming the IPsec mesh). As a motivating example, I've found *floating* IPv6 comparatively rare, for example, but we can _always_ support a dual-stack even uni-ipv6-stacked access address for a Postgres server inside the encapsulated network.

An important errata involves the dissatisfying double-encryption of services that implement TLS within IPsec: we could also enable non-TLS connections if we know for certain a network block is implemented using IPsec, but this also disables handy TLS-linked features as well, such as certificate authentication, so I'm not sure if we want to go there as a default option. Most database servers are low volume, and some of those are still important nevertheless, and any extra margin of safety and administrative simplification to that effect is appreciated. So I don't think we should get too wound up about double-enciphering until we have such a hot database server that we want to fast-path Postgres occasionally by eliminating TLS.

Summing up, what should the commit message say? This sprawling communication of mine is too disorganized and contextual. But I think a starting point is that we want to secure the two properties of shielding Postgres from the Internet (simple to justify) and obtaining flexibility over how it is addressed relative to its physical infrastructure, for numerous regions (harder to convey at the right level of detail).

**Review by furkansahin — APPROVED** [review]
